### PR TITLE
NIFI-4598: When we retrieve the 'controller' from a remote NiFi insta…

### PR DESCRIPTION
**Pull request by markap14**
…nce in order to determine which ports are available, cache those results for up to some configurable amount of time (default 30 secs) so that we don't constantly issue HTTP Requests to the remote nifi

Thank you for submitting a contribution to Apache NiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [ ] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
- [ ] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
- [ ] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by bbende** [comment]
Reviewing...

File: nifi-commons/nifi-site-to-site-client/src/main/java/org/apache/nifi/remote/client/CachingRemoteGroupContentsManager.java [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+import org.apache.nifi.remote.util.SiteToSiteRestApiClient;
+import org.apache.nifi.web.api.dto.ControllerDTO;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class CachingRemoteGroupContentsManager implements RemoteGroupContentsManager {
```
Comment (bbende) [inline nifi-commons/nifi-site-to-site-client/src/main/java/org/apache/nifi/remote/client/CachingRemoteGroupContentsManager.java]: It looks like this code is also in the API client so I'm assuming it should only exist in one place and one of them is probably left-over

Comment (markap14) [reply nifi-commons/nifi-site-to-site-client/src/main/java/org/apache/nifi/remote/client/CachingRemoteGroupContentsManager.java]: D'oh! Right you are. Started implementing it in a separate class, then realized that it worked a lot more nicely to just embed the logic in the client and forgot to remove the interface & impl. Will push another commit.

**Comment by bbende** [comment]
+1 verified the caching behavior is working as expected, will merge to master

**Pull request closed without merging**